### PR TITLE
PacketIO server implementation

### PR DESCRIPTION
**Pull request by DanG100**
This is a breaking change, so i'm putting behind a flag. I'll need modify lemming to support this new model

### CODE REVIEW

File: dataplane/forwarding/fwdport/ports/cpu.go [hunk 1]
```diff
@@ -167,8 +179,53 @@ func (p *cpuPort) punt(v interface{}) {
...
+	ps := p.ctx.CPUPortSink()
+	p.ctx.RUnlock()
+	if ps != nil {
+		timer := deadlock.NewTimer(deadlock.Timeout, fmt.Sprintf("Punting packet from port %v", p))
+		err := ps(response)
+		timer.Stop()
+		if err == nil {
+			return
+		}
+		log.Errorf("ports: Unable to punt packet, request %+v, err %v.", response, err)
```
Comment (guoshiuan) [inline dataplane/forwarding/fwdport/ports/cpu.go:222]: Should it be treated as tx error like when ps == nil?

File: dataplane/forwarding/fwd.go [hunk 2]
```diff
@@ -722,78 +724,87 @@ func (e *Server) FlowCounterQuery(_ context.Context, request *fwdpb.FlowCounterQ
...
+}
+
+// PacketInject injects a packet in the specified forwarding context and port.
+func (e *Server) PacketInject(srv fwdpb.Forwarding_PacketInjectServer) error {
+	for {
+		request, err := srv.Recv()
+		if err != nil {
+			return err
+		}
+		err = e.injectPacket(request.GetContextId(), request.GetPortId(), request.GetStartHeader(),
```
[the remark this answers is not in the document]
Comment (DanG100) [reply dataplane/forwarding/fwd.go:804]: Recv() is blocking, we want to RPC to remain running until is closed by the client

File: dataplane/forwarding/fwd.go [hunk 3]
```diff
@@ -722,78 +724,87 @@ func (e *Server) FlowCounterQuery(_ context.Context, request *fwdpb.FlowCounterQ
...
 				}
-				fwdport.Process(port, packet, request.GetAction(), ctx, "Control")
 			}()
+
+			packet.Debug(debug)
+			if len(pre) != 0 {
+				packet.Log().WithValues("context", ctx.ID, "port", port.ID())
+				state, err := fwdaction.ProcessPacket(packet, pre, port)
+				if state != fwdaction.CONTINUE || err != nil {
+					log.Errorf("%v: preprocessing failed, state %v, err %v", port.ID(), state, err)
```
[the remark this answers is not in the document]
Comment (DanG100) [reply dataplane/forwarding/fwd.go:782]: generally we don't care about error for individual packets, it's to the application layer to handle those

File: dataplane/forwarding/fwdport/ports/cpu.go [hunk 4]
```diff
@@ -167,8 +179,53 @@ func (p *cpuPort) punt(v interface{}) {
...
 
+func (p *CPUPort) puntRemotePort(v any) {
+	packet, ok := v.(fwdpacket.Packet)
+	if !ok {
+		fwdport.Increment(p, 1, fwdpb.CounterId_COUNTER_ID_TX_ERROR_PACKETS, fwdpb.CounterId_COUNTER_ID_TX_ERROR_OCTETS)
+		return
+	}
+
+	p.ctx.RLock()
+	var ingressPID *fwdpb.PortId
```
[the remark this answers is not in the document]
Comment (DanG100) [reply dataplane/forwarding/fwdport/ports/cpu.go:190]: yes, this just metadata, however the input port should also be set

File: dataplane/forwarding/fwdport/ports/cpu.go [hunk 5]
```diff
@@ -167,8 +179,53 @@ func (p *cpuPort) punt(v interface{}) {
...
+	if port, err := fwdport.InputPort(packet, p.ctx); err == nil {
+		ingressPID = fwdport.GetID(port)
+	}
+	egressPID := fwdport.GetID(p)
```
[the remark this answers is not in the document]
Comment (DanG100) [reply dataplane/forwarding/fwdport/ports/cpu.go:194]: yes

File: dataplane/forwarding/fwdport/ports/cpu.go (hunk 1 above)
Comment (DanG100) [reply dataplane/forwarding/fwdport/ports/cpu.go]: yes good catch, fixed

**Review by guoshiuan — APPROVED** [review]

**Pull request merged**